src/drivers/imx/sdma.c

```diff
@@ -0,0 +1,848 @@
+				   dma->plat_data.channels * sizeof(struct sdma_ccb));
+	if (!pdata->ccb_array) {
+		ret = -ENOMEM;
+		trace_sdma_error("SDMA: probe: unable to allocate CCBs");
+		goto err;
+	}
+
+	ret = sdma_boot(dma);
+	if (ret < 0) {
+		trace_sdma_error("SDMA: Unable to boot");
+		goto err;
+	}
+
+	goto out;
+err:
+	if (pdata->chan_pdata)
+		rfree(pdata->chan_pdata);
+	if (pdata->contexts)
+		rfree(pdata->contexts);
+	if (pdata->ccb_array)
+		rfree(pdata->ccb_array);
+	/* Failures of allocation were treated already */
+	rfree(dma_get_drvdata(dma));
+	rfree(dma->chan);
+	dma_set_drvdata(dma, NULL);
+	dma->chan = NULL;
+out:
+	trace_sdma_error("sdma_probe done, ret = %d", ret);
+	return ret;
+}
+
+static int sdma_remove(struct dma *dma) {
+	if (!dma->chan) {
+		trace_sdma_error("SDMA: Remove called without probe, that's a noop");
+		return 0;
+	}
+
+	trace_sdma_error("sdma_remove");
+
+	/* Prevent all channels except channel 0 from running */
+	dma_reg_write(dma, SDMA_HOSTOVR, 1);
+
+	/* Stop all channels except channel 0 */
+	dma_reg_write(dma, SDMA_STOP_STAT, ~1);
+
+	/* Tear down contexts */
+	sdma_teardown_contexts(dma);
+
+	/* Shut down controller */
+	sdma_shutdown(dma);
+
+	rfree(dma->chan);
+	dma->chan = NULL;
+
+	trace_sdma_error("sdma_remove done");
+	return 0;
+}
+
+static struct dma_chan_data *sdma_channel_get(struct dma *dma, unsigned chan) {
+	struct sdma_pdata *pdata = dma_get_drvdata(dma);
+	/* Ignoring channel; let's just allocate a free channel */
+
+	trace_sdma_error("sdma_channel_get");
+	for (int i = 0; i < dma->plat_data.channels; i++) {
+		struct dma_chan_data *channel = &dma->chan[i];
+		struct sdma_chan *cdata = &pdata->chan_pdata[i];
+
+		if (channel->status != COMP_STATE_INIT)
+			continue;
+
+		cdata->ctx = pdata->contexts + i;
+		cdata->ccb = pdata->ccb_array + i;
+
+		channel->status = COMP_STATE_READY;
+		dma_chan_set_data(channel, cdata);
+		/* Allow events, allow manual */
+		sdma_set_overrides(channel, false, true);
+		trace_sdma_error("sdma_channel_get returned channel %d", channel->index);
+		return channel;
+	}
+	trace_sdma_error("sdma no channel free");
+	return NULL;
+}
+
+static void sdma_clear_event(struct dma_chan_data *channel) {
+	struct sdma_chan *pdata = dma_chan_get_data(channel);
+
+	trace_sdma_error("sdma_clear_event(%d); old event is %d", channel->index, pdata->hw_event);
+	if (pdata->hw_event != -1)
+		dma_reg_update_bits(channel->dma, SDMA_CHNENBL(pdata->hw_event), BIT(channel->index), 0);
+	pdata->hw_event = -1;
+}
+
+static void sdma_set_event(struct dma_chan_data *channel, int eventnum) {
+	if (eventnum < -1 || eventnum > SDMA_HWEVENTS_COUNT)
+		return;
+	sdma_clear_event(channel);
+	if (eventnum == -1) {
+		sdma_set_overrides(channel, true, false);
+		return;
+	}
+	trace_sdma_error("sdma_set_event(%d, %d)", channel->index, eventnum);
+	struct sdma_chan *pdata = dma_chan_get_data(channel);
+	dma_reg_update_bits(channel->dma, SDMA_CHNENBL(eventnum), BIT(channel->index), BIT(channel->index));
```

Comment:
I would suggest on declaring all the variables at the beginning of function.

Reply:
Will do.